File: game/forest/talking_game_over.py
import pygame

from animation import Animation
from audio_helper import AudioHelper
from forest.forest_resources import ForestResources
from game_data import GameData
from null_state import NullState
from render_type import RenderType
from state import State


class TalkingGameOver(State):
    def __init__(self, context: GameData):
        super().__init__(context)
        self.needs_background = RenderType.PRE
        self.needs_bottom = RenderType.PRE

    def process_events(self, events):
        if self.get_frame_index() >= len(ForestResources.sync_gameover):
            return NullState
        return None

    def on_enter(self) -> None:
        super().on_enter()
        AudioHelper.play(ForestResources.speak_gameover, self.context.audio_port)

    def render(self, screen):
        screen.blit(Animation.get_sync_frame(ForestResources.hugo_telllives, ForestResources.sync_gameover, self.get_frame_index()), (128,-16))

File: game-web/src/forest/talkingGameOver.ts
/** Port of game/forest/talking_game_over.py */

import { State, type StateClass } from '../state';
import { Animation } from '../animation';
import { AudioHelper } from '../audioHelper';
import { ForestResources } from './forestResources';
import { NullState } from '../nullState';
import { RenderType } from '../renderType';
import type { GameData } from '../gameData';
import type { PhoneEvents } from '../phoneEvents';

export class TalkingGameOver extends State {
  readonly needsBackground = RenderType.PRE;
  readonly needsBottom = RenderType.PRE;

  constructor(context: GameData) { super(context); }

  processEvents(_events: PhoneEvents): StateClass | null {
    return this.getFrameIndex() >= ForestResources.syncGameover.length ? NullState : null;
  }

  onEnter(): void {
    super.onEnter();
    AudioHelper.play(ForestResources.speakGameover);
  }

  render(ctx: OffscreenCanvasRenderingContext2D): void {
    ctx.drawImage(Animation.getSyncFrame(ForestResources.hugoTelllives, ForestResources.syncGameover, this.getFrameIndex()), 128, -16);
  }
}
